application/core/matcher/graph.go

```diff
@@ -0,0 +1,132 @@
+// Package graph creates a ItemGraph data structure for the Item type
+package matcher
+
+import (
+    "fmt"
+    "sync"
+
+    //"github.com/cheekybits/genny/generic"
+)
+
+// Item the type of the binary search tree
+//type Item generic.Type
+
+// Node a single node that composes the tree
+type User struct {
+    userId string       
+}
+
+
+
+
+func (n *User) String() string {
+    return fmt.Sprintf("%s", n.userId)
+}
+
+// UserGraph the Items graph
+type UserGraph struct {
+    users []*User
+    edges map[User][]*User
+    lock  sync.RWMutex
+}
+
+// AddNode adds a node to the graph
+func (g *UserGraph) AddUser(n *User) {
+    g.lock.Lock()
+    g.users = append(g.users, n)
+    g.lock.Unlock()
+}
+
+// AddEdge adds an edge to the graph
+func (g *UserGraph) AddEdge(n1, n2 *User) {
+    g.lock.Lock()
+    if g.edges == nil {
+        g.edges = make(map[User][]*User)
+    }
+    g.edges[*n1] = append(g.edges[*n1], n2)
+    g.edges[*n2] = append(g.edges[*n2], n1)
+    g.lock.Unlock()
```

Comment:
@terrencetao can you move it right after the lock at line 42 and use `defer g.lock.Unlock()`